Source/CarthageKit/NewResolver.swift

```diff
@@ -0,0 +1,557 @@
+import Foundation
+import Result
+import ReactiveSwift
+
+/// Responsible for resolving acyclic dependency graphs.
+public struct NewResolver: ResolverProtocol {
+	private let versionsForDependency: (Dependency) -> SignalProducer<PinnedVersion, CarthageError>
+	private let resolvedGitReference: (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>
+	private let dependenciesForDependency: (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>
+
+	/// Instantiates a dependency graph resolver with the given behaviors.
+	///
+	/// versionsForDependency - Sends a stream of available versions for a
+	///                         dependency.
+	/// dependenciesForDependency - Loads the dependencies for a specific
+	///                             version of a dependency.
+	/// resolvedGitReference - Resolves an arbitrary Git reference to the
+	///                        latest object.
+	public init(
+		versionsForDependency: @escaping (Dependency) -> SignalProducer<PinnedVersion, CarthageError>,
+		dependenciesForDependency: @escaping (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>,
+		resolvedGitReference: @escaping (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>
+	) {
+		self.versionsForDependency = versionsForDependency
+		self.dependenciesForDependency = dependenciesForDependency
+		self.resolvedGitReference = resolvedGitReference
+	}
+
+	/// Attempts to determine the latest valid version to use for each
+	/// dependency in `dependencies`, and all nested dependencies thereof.
+	///
+	/// Sends a dictionary with each dependency and its resolved version.
+	public func resolve(
+		dependencies: [Dependency: VersionSpecifier],
+		lastResolved: [Dependency: PinnedVersion]? = nil,
+		dependenciesToUpdate: [String]? = nil
+		) -> SignalProducer<[Dependency: PinnedVersion], CarthageError> {
+		let result = process(dependencies: dependencies, in: DependencyGraph(whitelist: dependenciesToUpdate, lastResolved: lastResolved))
+			.map { graph -> [Dependency: PinnedVersion] in
+				guard
+					let dependenciesToUpdate = dependenciesToUpdate,
+					let lastResolved = lastResolved,
+					!dependenciesToUpdate.isEmpty else {
+						return graph.versions()
+				}
+
+				var newVersions = graph.versions(for: dependenciesToUpdate)
+				for (dependency, version) in lastResolved {
+					// If we don't have a new version for it, and it's still in the graph, grab the last resolved version
+					guard newVersions[dependency] == nil && graph.node(for: dependency) != nil else { continue }
+					newVersions[dependency] = version
+				}
+
+				return newVersions
```

Comment:
Actually, since I later added the fix for skipping updates that don't fit the whitelist,  it's even easier. We only need to filter out any new dependencies that didn't exist before in 'validateFinalGraph', then all the whitelist logic can be encapsulated there